addons/stock_picking_batch/models/stock_picking_batch.py

```diff
@@ -185,23 +185,23 @@ def action_print(self):
         return self.env.ref('stock_picking_batch.action_report_picking_batch').report_action(self)
 
     def action_done(self):
-        def has_no_quantity(picking):
-            return all(not m.picked or float_is_zero(m.quantity, precision_rounding=m.product_uom.rounding) for m in picking.move_ids if m.state not in ('done', 'cancel'))
+        def is_unpicked(picking):
+            return all(not m.picked for m in picking.move_ids if m.state not in ('done', 'cancel'))
 
-        def is_empty(picking):
+        def has_no_quantity(picking):
             return all(float_is_zero(m.quantity, precision_rounding=m.product_uom.rounding) for m in picking.move_ids if m.state not in ('done', 'cancel'))
 
         self.ensure_one()
         self._check_company()
         # Empty 'assigned' or 'waiting for another operation' pickings will be removed from the batch when it is validated.
         pickings = self.mapped('picking_ids').filtered(lambda picking: picking.state not in ('cancel', 'done'))
-        empty_waiting_pickings = self.mapped('picking_ids').filtered(lambda p: (p.state == 'waiting' and has_no_quantity(p)) or (p.state == 'assigned' and is_empty(p)))
+        empty_waiting_pickings = self.mapped('picking_ids').filtered(lambda p: p.state == 'waiting' and (has_no_quantity(p) or is_unpicked(p)))
         pickings = pickings - empty_waiting_pickings
 
-        empty_pickings = set()
+        empty_pickings = self.env['stock.picking']
```

Comment:
I prefer using an empty `stock.picking` object because it allows us to skip additional computations like converting to a list() and browsing, especially since we remove empty pickings in subsequent steps from the pickings